src/bitcoin.cpp

```diff
@@ -0,0 +1,207 @@
+            throw std::runtime_error(strprintf("Unknown option: %s", arg));
+        } else if (!arg.empty()) {
+            cmd.command = arg;
+        }
+    }
+    return cmd;
+}
+
+// Execute the specified bitcoind, bitcoin-qt or other command line in `args`
+// using src, bin and libexec directory paths relative to this executable, where
+// the path to this executable is specified `argv0`.
+//
+// This function doesn't currently print anything but can be debugged from
+// the command line using strace like:
+//
+//     strace -e trace=execve -s 10000 build/src/bitcoin ...
+void ExecCommand(const std::vector<std::string>& args, std::string_view argv0)
+{
+    // Construct argument string for execvp
+    std::vector<const char*> cstr_args{};
+    cstr_args.reserve(args.size() + 1);
+    for (const auto& arg : args) {
+        cstr_args.emplace_back(arg.c_str());
+    }
+    cstr_args.emplace_back(nullptr);
+
+    // Try to call execvp with given exe path.
+    auto try_exec = [&](fs::path exe_path, bool allow_notfound = true) {
+        std::string exe_path_str{fs::PathToString(exe_path)};
+        cstr_args[0] = exe_path_str.c_str();
+        if (execvp(cstr_args[0], (char*const*)cstr_args.data()) == -1) {
+            if (allow_notfound && errno == ENOENT) return false;
+            throw std::system_error(errno, std::system_category(), strprintf("execvp failed to execute '%s'", cstr_args[0]));
+        }
+        return true; // Will not actually be reached if execvp succeeds
+    };
+
+    // Try to figure out where current executable is located. This is a
+    // simplified search that won't work perfectly on every platform and doesn't
+    // need to, as it is only trying to prioritize locally built or installed
+    // executables over system executables. We may want to add options to
+    // override this behavior in the future, though.
+    const fs::path exe_path{fs::PathFromString(std::string{argv0})};
+    fs::path exe_dir(argv0);
+    std::error_code ec;
+#ifndef WIN32
+    if (argv0.find('/') == std::string_view::npos) {
+        if (const char* path_env = std::getenv("PATH")) {
+            size_t start{0}, end{0};
+            for (std::string_view paths{path_env}; end != std::string_view::npos; start = end + 1) {
+                end = paths.find(':', start);
+                fs::path candidate = fs::path(paths.substr(start, end - start)) / exe_dir;
+                if (fs::exists(candidate, ec) && fs::is_regular_file(candidate, ec)) {
+                    exe_dir = candidate;
+                    break;
+                }
+            }
+        }
+    }
+#else
+    wchar_t module_path[MAX_PATH];
+    if (GetModuleFileNameW(nullptr, module_path, MAX_PATH) > 0) {
+        exe_dir = fs::path{module_path};
+    } else {
+        tfm::format(std::cerr, "Warning: Failed to get executable path. Error: %s\n", GetLastError());
+    }
+#endif
+    exe_dir = fs::weakly_canonical(exe_dir, ec).parent_path();
```

Comment:
f17453ff13ebd0be842396c17bcc99b0c903abb8: shouldn't we throw here if `ec` is set?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31375#discussion_r1884865585

> [f17453f](https://github.com/bitcoin/bitcoin/commit/f17453ff13ebd0be842396c17bcc99b0c903abb8): shouldn't we throw here if `ec` is set?

I don't think we actually want to throw when symlinks can't be resolved. Should be more robust to just search for executables relative to the original path in that case. But error handling here was not right, because weakly_canonical would return an empty path on failure, so that should be fixed now.

Reply:
Ah, the new comment is helpful.